File: Projects/7. iExpense/iExpense/iExpense/ExpenseItem.swift
//
//  ExpenseItem.swift
//  iExpense
//
//  Created by Виталий Овсянников on 11.07.2024.
//

import Foundation

extension Expenses {
	struct Item: Codable, Identifiable, Hashable {
		private(set) var id = UUID()

		let name: String
		let type: ExpenseType
		let amount: Double
	}
}

extension Expenses.Item {
	enum ExpenseType: String, CaseIterable, Codable {
		case personal = "Персональный"
		case business = "Бизнес"
	}
}

File: Jetpack Compose/4.iExpense/app/src/main/java/com/example/iexpense/data/ExpenseItem.kt
package com.example.iexpense.data

import androidx.compose.runtime.Immutable
import kotlinx.serialization.Serializable
import java.text.NumberFormat
import java.util.Locale
import java.util.UUID

@Immutable
@Serializable
data class ExpenseItem(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val type: ExpenseType = ExpenseType.PERSONAL,
    val amount: Float
) {
    val formattedAmount: String
        get() {
            val locale = Locale.getDefault()
            val formatter = NumberFormat.getCurrencyInstance(locale)

            return formatter.format(amount)
        }

    val formattedType: String
        get() = type.displayName
}

enum class ExpenseType {
    PERSONAL, BUSINESS;

    val displayName = name.lowercase().replaceFirstChar { it.uppercase() }
}